src/main/scala-2.11/markets/clearing/engines/CDAMatchingEngine.scala

```diff
@@ -15,24 +15,208 @@ limitations under the License.
 */
 package markets.clearing.engines
 
-import java.util.UUID
-
-import markets.clearing.strategies.BestLimitPriceFormationStrategy
-import markets.orders.orderings.PriceOrdering
+import markets.clearing.engines.matches.{Match, PartialMatch, TotalMatch}
+import markets.orders.limit.{LimitBidOrder, LimitAskOrder, LimitOrderLike}
+import markets.orders.market.{MarketOrderLike, MarketAskOrder, MarketBidOrder}
 import markets.orders.{AskOrder, BidOrder, Order}
+import markets.orders.orderings.PriceOrdering
+
+import scala.annotation.tailrec
+import scala.collection.{immutable, mutable}
 
-import scala.collection.immutable
 
+/** Continuous Double Auction (CDA) Matching Engine. */
+class CDAMatchingEngine(askOrdering: PriceOrdering[AskOrder],
+                        bidOrdering: PriceOrdering[BidOrder],
+                        initialPrice: Long) extends MutableMatchingEngine {
 
-class CDAMatchingEngine(val askOrdering: PriceOrdering[AskOrder],
-                        val bidOrdering: PriceOrdering[BidOrder],
-                        initialPrice: Long)
-  extends CDAMatchingEngineLike
-  with BestLimitPriceFormationStrategy {
+  /* Mutable collection of ask orders for internal use only! */
+  protected val _askOrderBook = mutable.TreeSet.empty[AskOrder](askOrdering)
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [It's preferred to use immutable collections instead of mutable ones](https://www.codacy.com/app/davidrpugh/markets-sandbox/file/1519375750/issues/source?bid=2921361&fileBranchId=3008855#l34)